File: go/tree/avlTree/avl.go
package avl

import "container/list"

type AVLNode struct {
	height int
	Data   int
	Left   *AVLNode
	Right  *AVLNode
}

type AVLTree struct {
	root *AVLNode
}

/*
 * @param data int
 * @return *AVLNode
 */
func NewAVLNode(data int) *AVLNode {
	return &AVLNode{
		height: 0,
		Data:   data,
		Left:   nil,
		Right:  nil,
	}
}

/*
 * @param root *AVLNode
 * @return int
 */
func GetHeight(node *AVLNode) int {
	if node == nil {
		return -1
	}
	return node.height
}

func NewAvlTree() *AVLTree {
	return &AVLTree{
		root: nil,
	}
}

func getMaxHeight(root *AVLNode) int {
	if root == nil {
		return 0
	}
	return max(GetHeight(root.Left), GetHeight(root.Right))
}

//旋转操作

// 左旋转
func leftRotate(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	newRoot := node.Right
	node.Right = newRoot.Left
	newRoot.Left = node
	node.height = getMaxHeight(node)
	newRoot.height = getMaxHeight(newRoot)
	return newRoot
}

// 右旋转
func rightRotate(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	newRoot := node.Left
	node.Left = newRoot.Right
	newRoot.Right = node
	node.height = getMaxHeight(node)
	newRoot.height = getMaxHeight(newRoot)
	return newRoot
}

// 左右旋转
func leftRightRotate(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	node.Left = leftRotate(node.Left)
	return rightRotate(node)
}

// 右左旋转
func rightLeftRotate(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	node.Right = rightRotate(node.Right)
	return leftRotate(node)
}

// 再平衡函数
func reBalance(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	if GetHeight(node.Left)-GetHeight(node.Right) == 2 {
		if GetHeight(node.Left.Left)-GetHeight(node.Left.Right) == 1 {
			node = rightRotate(node)
		} else {
			node = leftRightRotate(node)
		}
	} else if GetHeight(node.Left)-GetHeight(node.Right) == -2 {
		if GetHeight(node.Right.Right)-GetHeight(node.Right.Left) == 1 {
			node = leftRotate(node)
		} else {
			node = rightLeftRotate(node)
		}
	}
	return node
}

// 插入
func (t *AVLTree) Insert(node *AVLNode, data int) *AVLNode {
	if node == nil {
		return NewAVLNode(data)
	}
	if data < node.Data {
		node.Left = t.Insert(node.Left, data)
	} else if data > node.Data {
		node.Right = t.Insert(node.Right, data)
	}
	node = reBalance(node)
	node.height = getMaxHeight(node) + 1
	return node
}

func (t *AVLTree) Delete(node *AVLNode, data int) *AVLNode {
	if node == nil {
		return nil
	}
	if data < node.Data {
		node.Left = t.Delete(node.Left, data)
	} else if data > node.Data {
		node.Right = t.Delete(node.Right, data)
	} else {
		if node.Left == nil {
			return node.Right
		} else if node.Right == nil {
			return node.Left
		} else {
			minNode := t.FindMin(node.Right)
			node.Data = minNode.Data
			node.Right = t.Delete(node.Right, minNode.Data)
		}
	}
	node = reBalance(node)
	node.height = getMaxHeight(node)
	return node
}

func (t *AVLTree) FindMin(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	if node.Left == nil {
		return node
	}
	return t.FindMin(node.Left)
}
func (t *AVLTree) FindMax(node *AVLNode) *AVLNode {
	if node == nil {
		return nil
	}
	if node.Right == nil {
		return node
	}
	return t.FindMax(node.Right)
}

func (t *AVLTree) InOrderArray() []int {
	arr := make([]int, 0)
	if t.root == nil {
		return arr
	}
	return inOrder(t.root, arr)
}

func inOrder(node *AVLNode, arr []int) []int {
	if node == nil {
		return arr
	}
	return append(append(inOrder(node.Left, arr), node.Data), inOrder(node.Right, arr)...)
}

func (t *AVLTree) LevelOrderArray() []int {
	return levelOrder(t.root)
}

func levelOrder(node *AVLNode) []int {
	arr := make([]int, 0)
	if node == nil {
		return arr
	}
	l := list.New()
	l.PushBack(node)
	for l.Len() > 0 {
		e := l.Front()
		l.Remove(e)
		node = e.Value.(*AVLNode)
		arr = append(arr, node.Data)
		if node.Left != nil {
			l.PushBack(node.Left)
		}
		if node.Right != nil {
			l.PushBack(node.Right)
		}
	}
	return arr
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: dart/src/tree/avlTree/avl.dart
import 'dart:collection';
import 'dart:math';

class AvlNode<T> {
  T? data;
  AvlNode<T>? left;
  AvlNode<T>? right;
  int height = 0;

  AvlNode(this.data, this.left, this.right);
}

class AvlTree<T extends Comparable> {
  AvlNode<T>? root;

  AvlTree() {
    this.root = null;
  }

  bool get isEmpty {
    return this.root == null;
  }

  int height(AvlNode<T>? node) {
    if (node == null) {
      return -1;
    }
    return node.height;
  }

  int getMaxHeight(AvlNode<T>? node) {
    if (node == null) {
      return -1;
    }
    int leftHeight = height(node.left);
    int rightHeight = height(node.right);
    return max(leftHeight, rightHeight) + 1;
  }

//   左旋转
  AvlNode<T>? leftRotate(AvlNode<T>? node) {
    AvlNode<T>? rightNode = node?.right;

    node?.right = rightNode?.left;
    rightNode?.left = node;

    node?.height = getMaxHeight(node);
    rightNode?.height = getMaxHeight(rightNode);
    return rightNode;
  }

//   右旋转
  AvlNode<T>? rightRotate(AvlNode<T>? node) {
    AvlNode<T>? leftNode = node?.left;

    node?.left = leftNode?.right;
    leftNode?.right = node;

    node?.height = getMaxHeight(node);
    leftNode?.height = getMaxHeight(leftNode);
    return leftNode;
  }

//   左右旋转
  AvlNode<T>? leftRightRotate(AvlNode<T>? node) {
    node?.left = leftRotate(node.left);
    return rightRotate(node);
  }

//   右左旋转
  AvlNode<T>? rightLeftRotate(AvlNode<T>? node) {
    node?.right = rightRotate(node.right);
    return leftRotate(node);
  }

  AvlNode<T>? insertNode(AvlNode<T>? node, T data) {
    if (node == null) {
      return AvlNode(data, null, null);
    }
    if (data.compareTo(node.data!) < 0) {
      node.left = insertNode(node.left, data);
    } else if (data.compareTo(node.data!) > 0) {
      node.right = insertNode(node.right, data);
    }
    node = balance(node);
    node!.height = getMaxHeight(node);
    return node;
  }

  AvlNode<T>? deleteNode(AvlNode<T>? node, T data) {
    if (node == null) {
      return null;
    }

    if (data.compareTo(node.data!) < 0) {
      node.left = deleteNode(node.left, data);
    } else if (data.compareTo(node.data!) > 0) {
      node.right = deleteNode(node.right, data);
    } else {
      if (node.left == null && node.right == null) {
        node = null;
      } else if (node.left == null) {
        node = node.right;
      } else if (node.right == null) {
        node = node.left;
      } else {
        var min = findMin(node.right);
        node.data = min;
        node.right = deleteNode(node.right, min!);
      }
    }
    node = balance(node);
    node?.height = getMaxHeight(node);
    return node;
  }

  AvlNode<T>? balance(AvlNode<T>? node) {
    if (node == null) {
      return null;
    }
    if (height(node.left) - height(node.right) > 1) {
      if (height(node.left?.left) >= height(node.left?.right)) {
        node = rightRotate(node);
      } else {
        node = leftRightRotate(node);
      }
    } else if (height(node.right) - height(node.left) > 1) {
      if (height(node.right?.right) >= height(node.right?.left)) {
        print("leftRotate");
        node = leftRotate(node);
      } else {
        node = rightLeftRotate(node);
      }
    }
    return node;
  }

  T? findMin(AvlNode<T>? node) {
    if (root == null) {
      return null;
    }
    while (node!.left != null) {
      node = node.left;
    }
    return node.data;
  }

  void delete(T data) {
    this.root = deleteNode(this.root, data);
  }

  void insert(T data) {
    this.root = insertNode(this.root, data);
  }

  /// Performs an in-order traversal of the AVL tree starting from the given node.
  ///
  /// The in-order traversal visits the nodes in ascending order based on their data.
  ///
  /// Parameters:
  /// - `node`: The starting node of the traversal. If `null`, the traversal is not performed.
  ///
  /// Returns:
  /// This function does not return any value.

  void inOrder(AvlNode<T>? node) {
    if (node == null) {
      return;
    }
    inOrder(node.left);
    print("${node.data} ${node.height}");
    inOrder(node.right);
  }

  List toList() {
    List list = [];
    if (root != null) {
      Queue<AvlNode<T>> queue = Queue();
      queue.addLast(root!);
      while (!queue.isEmpty) {
        AvlNode<T> current = queue.removeFirst();
        list.add(current.data);
        if (current.left != null) {
          queue.addLast(current.left!);
        }
        if (current.right != null) {
          queue.addLast(current.right!);
        }
      }
    }
    return list;
  }
}
